src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs

```diff
@@ -2,25 +2,54 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Headers;
+using Microsoft.AspNetCore.Internal;
 using Microsoft.Extensions.Logging;
 using Microsoft.Net.Http.Headers;
 
 namespace Microsoft.AspNetCore.Mvc.Internal
 {
     public class FileResultExecutorBase
     {
+        private const string AcceptRangeHeaderValue = "bytes";
+
         public FileResultExecutorBase(ILogger logger)
         {
             Logger = logger;
         }
 
         protected ILogger Logger { get; }
 
-        protected void SetHeadersAndLog(ActionContext context, FileResult result)
+        protected (RangeItemHeaderValue range, long rangeLength)? SetHeadersAndLog(ActionContext context, FileResult result, long fileLength, DateTimeOffset? lastModified = null, EntityTagHeaderValue etag = null)
         {
             SetContentType(context, result);
             SetContentDispositionHeader(context, result);
             Logger.FileResultExecuting(result.FileDownloadName);
+            SetAcceptRangeHeader(context);
+            var httpRequestHeaders = context.HttpContext.Request.GetTypedHeaders();
+            if (context.HttpContext.Request.Headers.ContainsKey(HeaderNames.Range))
+            {
+                bool shouldProcess = ComputeConditionalRequestHeaders(
```

Comment:
This method ComputeConditionalRequestHeaders does too much, it not just only computes the headers but in some cases it also writes the response. I would suggest you factor this in a different way.
Have a method that tells you what to do, e.g. tells you if you should process the request headers or if you should send back precondition failed or other type of response.
Then based on the result do what you need to do (set the headers, send a precondition failed response, or similar).